### PR TITLE
[Workflow] CS tweaks

### PR DESCRIPTION
**Pull request by ro0NL**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | ~ |

Just a few minor tweaks/fixes after playing around with it. Some phpdoc should still be added imo, like missing `@throws`, right?

### CODE REVIEW

File: src/Symfony/Component/Workflow/Workflow.php [hunk 1]
```diff
@@ -59,10 +59,6 @@ public function getMarking($subject)
     {
         $marking = $this->markingStore->getMarking($subject);
 
-        if (!$marking instanceof Marking) {
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Workflow.php]: This should stop imo :-)

File: src/Symfony/Component/Workflow/Workflow.php [hunk 2]
```diff
@@ -59,10 +59,6 @@ public function getMarking($subject)
...
-            throw new LogicException(sprintf('The value returned by the MarkingStore is not an instance of "%s" for workflow "%s".', Marking::class, $this->name));
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Workflow.php]: Otherwise, technically this should be `\UnexpectedValueException`

File: src/Symfony/Component/Workflow/Definition.php [hunk 3]
```diff
@@ -68,7 +66,7 @@ public function addPlace($place)
             throw new InvalidArgumentException(sprintf('The place "%s" contains invalid characters.', $place));
         }
 
-        if (!count($this->places)) {
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Definition.php]: This is up for [debate](https://github.com/symfony/symfony/issues/19182) and maybe out of scope here..

File: src/Symfony/Component/Workflow/Workflow.php (hunk 1 above)
Comment (lyrixx) [inline src/Symfony/Component/Workflow/Workflow.php]: Why did you removed this code ?

Comment (ro0NL) [inline src/Symfony/Component/Workflow/Workflow.php]: https://github.com/ro0NL/symfony/blob/7bbf948d2499c0d5dfa5a0259cd3f9bbc3ea81d6/src/Symfony/Component/Workflow/MarkingStore/MarkingStoreInterface.php#L28

Comment (HeahDude) [inline src/Symfony/Component/Workflow/Workflow.php]: I agree this should not be needed though. Otherwise it would mean that the marking store is violating the interface contract made [here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/MarkingStore/MarkingStoreInterface.php#L28) and [here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Workflow/Workflow.php#L32). Does that make sense ?

Comment (lyrixx) [inline src/Symfony/Component/Workflow/Workflow.php]: I'm :-1: with this change. Nothing in PHP 5 forbid someone to violate the return type of an interface.

Comment (ro0NL) [inline src/Symfony/Component/Workflow/Workflow.php]: The violation should be fixed, not every case where it could/will cause a problem.

Comment (lyrixx) [inline src/Symfony/Component/Workflow/Workflow.php]: Yes it should be fixed. That's why there was an exception to ease debug. I don't understand why you want to remove that.

Comment (dosten) [inline src/Symfony/Component/Workflow/Workflow.php]: Actually there are a lot of places in the Symfony codebase where we skip checks like this because the interface define the expected type of argument, @fabpot can you argue over this?

Comment (nicolas-grekas) [inline src/Symfony/Component/Workflow/Workflow.php]: This is a DX issue. Having an explicit exception is way better than waiting for some random side effect to break the interface in case someone misses the contract. If we need a criteria to decide why the check is desired here but not in other places, I'd say: is it common for users to write their own implementation of the interface ? Then let's do the check. Otherwise it's mostly Symfony internal and no check is fine.

Comment (ro0NL) [inline src/Symfony/Component/Workflow/Workflow.php]: I guess we're not gonna agree on this. However i understand the concerns when it comes to "missing the contract and debugging", but the exception would be cosmetic only. Ie. the plain error you get 9/10 is just as informative (stacktrace wise), eventually you'll have to debug your implementation anyway. And if error handling is poor, i would advice fixing that.

> is it common for users to write their own implementation of the interface

From a symfony/framework pov this is almost any interface... which is exactly why i think we have to draw the line on this.

For debugging/validating on return types; i think this should be be mostly covered by static code analysis, tests, and eventually PHP7.

This is way i changed it. :)

**Comment by fabpot** [comment]
I'm against adding `@throws` params except on interfaces when the contract is defined. They aren't that useful and are a nightmare to maintain especially if you start listing all possible exceptions including the ones coming from called code. And if you are not listing all of them, then it's not that useful anymore.

File: src/Symfony/Component/Workflow/Workflow.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Component/Workflow/Workflow.php]: I agree with @nicolas-grekas here. We don't have that many interfaces in Symfony that should almost *always* be implemented by the end user.

**Comment by ro0NL** [comment]
@fabpot  do you find the custom SPL exceptions useful anyway? `Symfony\..\LogicException` vs. `\LogicException`. For type/domain checking i would be perfectly fine with the latter. Ie. imo custom exceptions only make sense when providing granularity (`InvalidSomethingException` for logic and `SomethingNotFound(::getSomething)` for runtime.

File: src/Symfony/Component/Workflow/Workflow.php (hunk 1 above)
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Workflow.php]: I think this interface is no different then most others. Its a pillar interface in the component with prepackaged/generic implemenations which will gonna be used 9/10.

DX-wise, the real wtf happens when PropertyAccess fails from these marking stores. You'll get an exception thats hard to relate to what you're probably doing. I also see no real added value to try/catch those and throw a `WorkflowException` from a previous exception, but that at least would be a true DX improvement imho.

Like i said, we're not gonna agree on this. So if symfony team votes >50% i'll remove/keep it then. Or however it's decided normally :)

Comment (nicolas-grekas) [inline src/Symfony/Component/Workflow/Workflow.php]: @ro0NL there is no downside into keeping this, please revert this change, let's move on. We won't have a vote on this but on the PR as a whole, as always.

**Comment by lyrixx** [comment]
@ro0NL custom exceptions made easier global catching. It's done like that in every other components.

**Comment by ro0NL** [comment]
Thats not the point ;-) A exception pointer interface per component is just fine, but do we need custom _SPL_ exceptions all the time, ie. an `InvalidArgumentException` per component. Or sometimes yes, sometimes no.

I see no added value in globally catching logic errors (ie programmers faults), however currently logic is thrown where its not necessarily a programmers fault. Anyway, totally out-of-scope, and guess also not something were gonna agree on :-)

File: src/Symfony/Component/Workflow/Workflow.php (hunk 1 above)
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Workflow.php]: OK. Im still not convinced though. Anyway hope to get some more feedback/thumbs this day then, and i'll have a look tonight. However i feel my arguments are not taken seriously.

Dont get me wrong;
> on the PR as a whole

From this debate on there's no clear yes/no till now, if you ask me, besides two :+1: for the the PR. Anyway, to move on, ill revert it from a "out of scope of this PR" point of view.

**Comment by ro0NL** [comment]
Reverted the the `count` and `instanceof` changes as it seems very opinionated, which i didnt expected to be honest... from my pov the latter was _the_ most important change code-wise.

Anyway, if this decision was rational, i guess its good to go :+1:

File: src/Symfony/Component/Workflow/Workflow.php [hunk 4]
```diff
@@ -59,6 +59,10 @@ public function getMarking($subject)
     {
         $marking = $this->markingStore->getMarking($subject);
 
+        if (!$marking instanceof Marking) {
+            throw new LogicException(sprintf('The value returned by the MarkingStore is not an instance of "%s" for workflow "%s".', Marking::class, $this->name));
```
Comment (ro0NL) [inline src/Symfony/Component/Workflow/Workflow.php]: For being a contract check, i guess `LogicException` is just fine also :) (opposed to `UnexpectedValue`)

**Comment by fabpot** [comment]
👍

**Comment by lyrixx** [comment]
:+1:

**Comment by ro0NL** [comment]
:+1: :)

**Comment by fabpot** [comment]
Thank you @ro0NL.
